File: android/app/src/main/kotlin/dev/deliteai/assistant/presentation/viewmodels/ChatViewModel.kt
/*
 * SPDX-FileCopyrightText: (C) 2025 DeliteAI Authors
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package dev.deliteai.assistant.presentation.viewmodels

import android.app.Application
import android.util.Log
import android.widget.Toast
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import dev.deliteai.NimbleNet
import dev.deliteai.assistant.domain.features.asr.ASRService
import dev.deliteai.assistant.domain.models.Chat
import dev.deliteai.assistant.domain.models.ChatMessage
import dev.deliteai.assistant.domain.repositories.CacheRepository
import dev.deliteai.assistant.domain.repositories.ChatRepository
import dev.deliteai.assistant.domain.repositories.LoggingRepository
import dev.deliteai.assistant.utils.Constants
import dev.deliteai.assistant.utils.ExceptionLogger
import dev.deliteai.assistant.utils.LoaderTextProvider
import dev.deliteai.assistant.utils.TAG
import dev.deliteai.assistant.utils.copyTextToClipboard
import dev.deliteai.datamodels.NimbleNetTensor
import dev.deliteai.impl.common.DATATYPE
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.buffer
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.filterIsInstance
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.util.Date
import java.util.UUID
import kotlin.math.min

class ChatViewModel(private val application: Application) : AndroidViewModel(application) {
    private val chatRepository = ChatRepository()
    private val cacheRepository = CacheRepository(application)
    private val loggingRepository = LoggingRepository()

    var outputStream = mutableStateOf<String?>(null)
    var currentTPS = mutableStateOf<Float?>(null)
    var chatHistory = mutableStateOf<List<ChatMessage>>(emptyList())
    var currentMessageLoading = mutableStateOf(true)
    var isChatScreenLoading = mutableStateOf(false)
    var isHistoryLoadInProgress = mutableStateOf(false)
    var currentWaitText = mutableStateOf("Loading...")
    private var currentChatId = mutableStateOf<String?>(null)
    var longTapMenuMessage = mutableStateOf<ChatMessage?>(null)

    // Voice overlay state
    var isOverlayVisible = mutableStateOf(false)
    var isUserSpeaking = mutableStateOf(false)

    // For voice recognition text handling
    var spokenText = mutableStateOf("")
    var persistedRecognizedText = mutableStateOf("")
    var volumeState = mutableFloatStateOf(0f)
    var audioPlaybackStatus = mutableStateOf(false)

    private var chattingJob: Job? = null
    private val addToChatMessageMutex = Mutex()
    var isInterruptButtonVisible = mutableStateOf(false)

    var isFirstMessageSent = mutableStateOf(false)

    private var waitMessageRefreshJob: Job? = null
    var topBarTitle = mutableStateOf<String?>(null)

    var thinkingStream = mutableStateOf<String>("")
    var masterOutputHolder = ""

    init {
        viewModelScope.launch(Dispatchers.IO) {
            chatRepository.getAudioPlaybackSignal().collect { isAudioPlaying ->
                Log.d(
                    TAG,
                    "monitorInterruptButtonVisibility: isAudioPlaying: $isAudioPlaying isLLM not Active: ${chattingJob?.isActive == false}"
                )
                audioPlaybackStatus.value = isAudioPlaying
                if (!isAudioPlaying && chattingJob?.isCompleted == true && chattingJob?.isCancelled == false && isOverlayVisible.value) {
                    delay(1000)
                    getLLMAudioFromVoiceInput()
                    withContext(Dispatchers.Main) {
                        isInterruptButtonVisible.value = false
                    }
                }
            }
        }
    }

    fun fetchModelName() {
        viewModelScope.launch(Dispatchers.IO) {
            val name = topBarTitle.value ?: chatRepository.getModelName()

            withContext(Dispatchers.Main) {
                topBarTitle.value = name
            }
        }
    }

    fun handleMessageLongTapAction(action: Constants.MESSAGE_LONG_TAP_ACTIONS) {
        if (longTapMenuMessage.value == null) return

        when (action) {
            Constants.MESSAGE_LONG_TAP_ACTIONS.COPY -> {
                copyToClipboard(longTapMenuMessage.value!!.message)
            }

            Constants.MESSAGE_LONG_TAP_ACTIONS.FLAG -> {
                flagMessage(longTapMenuMessage.value!!.message)
            }
        }

        longTapMenuMessage.value = null
    }

    private fun flagMessage(message: String?) {
        Toast.makeText(application, "Thanks for providing feedback", Toast.LENGTH_SHORT).show()

        viewModelScope.launch(Dispatchers.IO) {
            runCatching {
                loggingRepository.flagLLMMessage(application, message)
            }.onFailure { e ->
                Log.e(TAG, "Failed to flag message", e)
            }
        }
    }

    private fun copyToClipboard(message: String?) {
        application.copyTextToClipboard(message)
        Toast.makeText(application, "Copied to clipboard", Toast.LENGTH_SHORT).show()
    }

    fun isChattingJobActive() = chattingJob?.isActive == true

    fun cancelLLMAndClearAudioQueue() {
        chattingJob?.cancel()
        chatRepository.resetAudioPlayer()
    }

    fun handleBack() {
        cancelLLMAndClearAudioQueue()
        isOverlayVisible.value = false
    }

    fun handleTextViewButtonClick(userInput: String) {
        if (isOverlayVisible.value) return

        if (chattingJob?.isActive == true) {
            cancelLLMAndClearAudioQueue()
        } else {
            if (userInput.any { it.isLetterOrDigit() }) {
                addNewMessageToChatHistory(userInput, true)
                getLLMTextFromTextInput(userInput)

            } else {
                Toast.makeText(
                    application,
                    "Please enter a message to continue",
                    Toast.LENGTH_SHORT
                )
                    .show()
            }
        }
    }

    private fun keepUpdatingWaitMessage() {
        val loaderTextProvider = LoaderTextProvider()
        val initialDelay = 1200L
        var delayOffset = 0L

        waitMessageRefreshJob = viewModelScope.launch(Dispatchers.Default) {
            while (true) {
                withContext(Dispatchers.Main) {
                    currentWaitText.value = loaderTextProvider.getLoaderText()
                }

                delay(initialDelay + delayOffset)
                delayOffset += initialDelay
            }
        }
    }

    fun shouldRestartListening() {
        if (!audioPlaybackStatus.value && chattingJob?.isCompleted != false) {
            getLLMAudioFromVoiceInput()
        }
    }

    fun getLLMAudioFromVoiceInput() {
        cancelLLMAndClearAudioQueue()
        if (!isFirstMessageSent.value) {
            isFirstMessageSent.value = true
        }
        spokenText.value = ""
        isInterruptButtonVisible.value = false
        persistedRecognizedText.value = ""
        val chatJob = viewModelScope.launch(
            Dispatchers.Default + CoroutineExceptionHandler { _, e ->
                handleException(e)
            }) {
            chatRepository.resetAudioPlayer()
            val response = ASRService.startAndroidListener()
                .onEach { response ->
                    volumeState.floatValue = response.volume
                    isUserSpeaking.value = response.isSpeaking
                }.filter { it.text.isNotBlank() }
                .onEach { spokenText.value = it.text }
                .filter { !it.isSpeaking && !it.isLoading }
                .onEach { persistedRecognizedText.value = it.text }
                .flowOn(Dispatchers.Main)
                .onEach { res ->
                    addToChatMessageMutex.withLock {
                        addNewMessageToChatHistory(res.text, true)
                    }
                }
                .firstOrNull()

            response?.text?.let {
                withContext(Dispatchers.Main) {
                    isInterruptButtonVisible.value = true
                }
                getLLMAudioResponse(it)
            }
        }
        chatJob.invokeOnCompletion { cause ->
            cause?.let { handleException(cause) }
            isInterruptButtonVisible.value = false
        }
        chattingJob = chatJob
    }

    fun getLLMTextFromTextInput(textInput: String) {
        thinkingStream.value = ""

        cancelLLMAndClearAudioQueue()
        if (!isFirstMessageSent.value) {
            isFirstMessageSent.value = true
        }

        currentMessageLoading.value = true
        if (chattingJob?.isActive != true) {
            keepUpdatingWaitMessage()
        }
        if (chattingJob?.isActive == true) {
            addNewMessageToChatHistory(outputStream.value.toString(), false)
            chattingJob?.cancel()
        }
        outputStream.value = ""
        val chatJob = viewModelScope.launch(
            Dispatchers.Default + CoroutineExceptionHandler { _, e ->
                handleException(e)
            }) {
            chatRepository.getLLMText(textInput)
                .onEach { result ->
                    handleLLMResult(result, false)
                }
                .flowOn(Dispatchers.Main)
                .filterIsInstance<ChatRepository.GenerateResponseJobStatus.Finished>()
                .first()
        }
        chatJob.invokeOnCompletion { cause ->
            cause?.let { handleException(cause) }
            isInterruptButtonVisible.value = false
        }
        chattingJob = chatJob
    }

    private suspend fun getLLMAudioResponse(textInput: String) = withContext(Dispatchers.Main) {
        Log.d(TAG, "LLM Triggered by $textInput")
        currentMessageLoading.value = true
        outputStream.value = ""
        chatRepository.getLLMAudio(textInput)
            .buffer()
            .onEach { result ->
                handleLLMResult(result, true)
            }
            .filterIsInstance<ChatRepository.GenerateResponseJobStatus.Finished>()
            .first()
    }

    fun addNewMessageToChatHistory(message: String, isUserInput: Boolean) {
        if (message == "null") return
        val newMessage = ChatMessage(
            message = message,
            isUserMessage = isUserInput,
            tps = currentTPS.value,
            timestamp = Date()
        )
        chatHistory.value += newMessage
        currentTPS.value = null
        saveChatToRepository()
    }

    private  val THINK_SENTINEL = "/think"

    // Fast non-overlapping counter
    private fun String.countOccurrences(needle: String): Int {
        if (needle.isEmpty()) return 0
        var count = 0
        var from = 0
        while (true) {
            val i = indexOf(needle, startIndex = from)
            if (i == -1) break
            count++
            from = i + needle.length
        }
        return count
    }

    private fun cleanForUi(s: String): String {
        return s.replace("<think>", "")
            .replace("</think>", "")
            .replace("<tool_call>", "-> EXECUTING TOOL CALL")
            .replace("</tool_call>", "")
            .replace("<|im_end|>", "")
            // optional: hide the sentinel from the UI as well
            .replace(THINK_SENTINEL, "")
    }

    private var thinkSeen = 0
    private var cutAfterThirdThink = -1


    private fun handleLLMResult(
        result: ChatRepository.GenerateResponseJobStatus,
        isAudioExpected: Boolean
    ) {
        when (result) {
            is ChatRepository.GenerateResponseJobStatus.GenerationInProgress -> {
                if (isAudioExpected)
                    currentMessageLoading.value = false
            }

            is ChatRepository.GenerateResponseJobStatus.Finished -> {
//                val finalOutput = outputStream.value.toString()
//                addNewMessageToChatHistory(finalOutput, false)
//                outputStream.value = null
//                isInterruptButtonVisible.value = false
            }

            is ChatRepository.GenerateResponseJobStatus.NextItem -> {
                val newChunk = result.outputText

                masterOutputHolder += newChunk

                val totalThinks = masterOutputHolder.countOccurrences(THINK_SENTINEL)

                if (totalThinks > thinkSeen) {
                    if (thinkSeen < 3 && totalThinks >= 3 && cutAfterThirdThink < 0) {
                        val thirdStart = masterOutputHolder.lastIndexOf(THINK_SENTINEL)
                        if (thirdStart != -1) {
                            cutAfterThirdThink = thirdStart + THINK_SENTINEL.length
                        }
                    }
                    thinkSeen = totalThinks
                    thinkingStream.value = "" // reset the "thinking" buffer at sentinel boundaries
                }

                if (cutAfterThirdThink >= 0) {
                    val visible = masterOutputHolder.substring(cutAfterThirdThink)
                    val cleaned = cleanForUi(visible)
                    outputStream.value = cleaned.substring(0,min(cleaned.length, 147))
                } else {
                    thinkingStream.value = cleanForUi(thinkingStream.value + newChunk)
                }

                if (!isAudioExpected && newChunk.contains(Regex("[A-Za-z0-9]"))) {
                    currentMessageLoading.value = false
                    waitMessageRefreshJob?.cancel()
                }

            }
        }
    }

    private fun saveChatToRepository() {
        currentChatId.value?.let { chatId ->
            viewModelScope.launch(Dispatchers.IO) {
                cacheRepository.cacheChat(Chat(id = chatId, messages = chatHistory.value))
            }
        }
    }

    fun clearContextAndStartNewChat() {
        isChatScreenLoading.value = true
        cancelLLMAndClearAudioQueue()

        viewModelScope.launch(
            Dispatchers.Default + CoroutineExceptionHandler { _, e ->
                ExceptionLogger.log("clear_context_start_new_chat", e)
            }) {
            val res = NimbleNet.runMethod("clear_prompt", inputs = hashMapOf())
            if (!res.status) {
                ExceptionLogger.log("clear_prompt", Throwable("clear prompt method failed"))
                return@launch
            }

            withContext(Dispatchers.Main) {
                chatHistory.value = emptyList()
                isFirstMessageSent.value = false
                createNewUUID()
            }
        }.invokeOnCompletion {
            outputStream.value = null
            isChatScreenLoading.value = false
        }
    }

    fun loadChatFromId(chatId: String) {
        isHistoryLoadInProgress.value = true
        chatHistory.value = mutableListOf()
        outputStream.value = null
        isFirstMessageSent.value = false

        viewModelScope.launch(
            Dispatchers.Default + CoroutineExceptionHandler { _, e ->
                clearContextAndStartNewChat()
            }) {
            val chat = cacheRepository.retrieveChat(chatId)
            currentChatId.value = chat.id
            chatHistory.value = chat.messages

            val historicalContext = JSONArray()
            chat.messages.forEach {
                historicalContext.put(
                    JSONObject(
                        mapOf(
                            "type" to (if (it.isUserMessage) "user" else "assistant"),
                            "message" to it.message
                        )
                    )
                )
            }

            val res = NimbleNet.runMethod(
                "set_context", hashMapOf(
                    "context" to NimbleNetTensor(
                        data = historicalContext,
                        datatype = DATATYPE.JSON_ARRAY,
                        shape = intArrayOf(historicalContext.length())
                    )
                )
            )
            check(res.status)
            if (chat.messages.isNotEmpty())
                isFirstMessageSent.value = true
        }.invokeOnCompletion {
            isHistoryLoadInProgress.value = false
        }
    }

    private fun createNewUUID() {
        currentChatId.value = UUID.randomUUID().toString()
    }

    override fun onCleared() {
        super.onCleared()
        cancelLLMAndClearAudioQueue()
    }

    private fun handleException(exception: Throwable) {
        when (exception) {
            is CancellationException ->
                viewModelScope.launch(Dispatchers.IO) {
                    try {
                        chatRepository.stopLLM()
                    } catch (e: Exception) {
                        ExceptionLogger.log("handleException", e)
                    }

                    currentMessageLoading.value = false
                    if (outputStream.value != null) {
                        val output = outputStream.value.toString()
                        if (output.isNotBlank())
                            addNewMessageToChatHistory(output, false)
                    }
                    outputStream.value = null
                }

            is SecurityException ->
                viewModelScope.launch(Dispatchers.Main) {
                    Toast.makeText(application, exception.message, Toast.LENGTH_SHORT)
                        .show()
                }

            else ->
                ExceptionLogger.log("handleException", exception)
        }
    }
}

File: ios/NimbleEdgeAI/presentation/ViewModel/ChatViewModel.swift
/*
 * SPDX-FileCopyrightText: (C) 2025 DeliteAI Authors
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import Foundation
import SwiftUI
import Combine
import NimbleNetiOS

class ChatViewModel: ObservableObject {
    private let chatRepository: ChatRepository = ChatRepository()
    private let cacheRepository: CacheRepository = CacheRepository()
    @Published var outputStream: String?
    @Published var currentTPS: Float?
    @Published var chatHistory: [ChatMessage] = []
    @Published var currentMessageLoading: Bool = true
    var currentChatId: String?
    var isVoiceMode: Bool?
    
    // Voice overlay state
    @Published var isOverlayVisible: Bool = false
    @Published var isInterruptButtonVisible:Bool = false

    
    @Published var isLLMActive: Bool = false
    @Published var isASRActive: Bool = false
    @Published var isUserSpeaking: Bool = false
    @Published var isLLMSpeaking: Bool = false
    
    // For voice recognition text handling
    @Published var spokenText: String = ""
    @Published var persistedRecognizedText: String = ""
    @Published var volumeState: Float = 0.0
    
    private var chattingTask: Task<Void, Never>? = nil
    private let addToChatMessageLock = NSLock()
    
    var onNewMessageAdded: ((_ message: ChatMessage) -> Void)?
    var onChatHistoryUpdated: ((_ chatHistory: [ChatMessage]) -> Void)?
    var onOutputStreamUpdated: ((_ outputStream: String?) -> Void)?
    var onLLMActive: ((_ isActive: Bool) -> Void)?
    
    private var cancellables = Set<AnyCancellable>()
    
    func setOverlayVisibility(_ visible: Bool) {
        self.isOverlayVisible = visible
    }
    func cancelTTS() {
        chatRepository.continuousAudioPlayer.stopAndResetPlayback()
        isLLMSpeaking = false
    }
    
    func set(chatID: String?) {
        if let chatID = chatID{
            self.currentChatId = chatID
        } else {
            createChatID()
        }
        retriveChatHistory()
    }
    
    func retriveChatHistory() {
        DispatchQueue.global().async(execute: { [weak self] in
            let chatHistory = self?.cacheRepository.retrieveChat(id: self?.currentChatId ?? "")?.messages ?? []
            DispatchQueue.main.async {
                self?.chatHistory = chatHistory
                self?.onChatHistoryUpdated?(chatHistory)
            }
        })
    }

    func cancelLLM(){
        chattingTask?.cancel()
    }
    
    func interruptResponse(){
        self.cancelLLM()
        self.cancelCurrentLLM()
        self.saveChatToRepository()
    }

    func createChatID() {
        currentChatId = UUID().uuidString
    }
    
    
    func cancelCurrentLLM() {
        if isLLMActive {
            if let outputStream = outputStream, !outputStream.isEmpty {
                self.saveChatToRepository()
            }
            DispatchQueue.global().async { [weak self] in
                self?.chatRepository.stopLLM()
            }
            
            isLLMActive = false
            DispatchQueue.main.async { [weak self] in
                self?.outputStream = nil
            }
            
        }
    }
    
    func clearContextAndStartNewChat(){
        
    }
    
    func passTextInputToLLM(_ textInput: String) {
        
        if(textInput == "") { return }
        
        var index = AtomicInteger(value: 1)
        DispatchQueue.main.async {
            self.currentMessageLoading = true
            self.outputStream = ""
            self.isLLMActive = true
            self.onLLMActive?(self.isLLMActive)
            let chatMessage = ChatMessage(message: "", isUserMessage: false, timestamp: Date(), tps: 0)
            self.chatHistory.append(chatMessage)
            self.onNewMessageAdded?(self.chatHistory.last!)
        }
        
        if isOverlayVisible {
            DispatchQueue.main.async { self.isLLMSpeaking = true }
            chattingTask = Task(priority: .low) {
                do {
                    await chatRepository.processUserInput(
                        textInput: textInput,
                        onOutputString: { [weak self] output in
                            if output.isEmpty { return }
                            DispatchQueue.main.async {
                                guard let self, self.isLLMActive else { return }
                                
                                
                                if let outputStream = self.outputStream, outputStream.isEmpty {
                                    //triming extra spcaes
                                    let newOutputStream = ((self.outputStream ?? "") + output)
                                    if let range = newOutputStream.range(of: "\\S", options: .regularExpression) {
                                        self.outputStream = newOutputStream[range.lowerBound...].description
                                        self.chatHistory[self.chatHistory.count - 1].message = self.outputStream
                                        self.onOutputStreamUpdated?(self.outputStream)
                                    }
                                } else {
                                    self.outputStream = (self.outputStream ?? "") + output
                                    self.chatHistory[self.chatHistory.count - 1].message = self.outputStream
                                    self.onOutputStreamUpdated?(self.outputStream)
                                }
                                
                            }
                        },
                        onFirstAudioGenerated: { [weak self] in
                            DispatchQueue.main.async {
                                self?.currentMessageLoading = false
                            }
                        },
                        onFinished: { [weak self] in
                            guard let self = self else { return }
                            
                            self.cancelLLM()
                            DispatchQueue.main.async {
                                self.cancelCurrentLLM()
                                self.isInterruptButtonVisible = false
                                self.isASRActive = true
                            }
                        },
                        onError: { [weak self] error in
                            print("LLM processing error: \(error)")
                            DispatchQueue.main.async {
                                self?.isLLMActive = false
                            }
                        },
                        onAudioFinishedPlaying: { [weak self] in
                            DispatchQueue.main.async {
                                self?.isLLMSpeaking = false
                            }
                            
                        }
                    )
                }
            }
        }
        else{
            chattingTask = Task(priority: .low) {
                do {
                    await chatRepository.processUserInput(
                        textInput: textInput,
                        onOutputString: { [weak self] output in
                            if output.isEmpty { return }
                            DispatchQueue.main.async {
                                guard let self else { return }
                                
                                if let outputStream = self.outputStream, outputStream.isEmpty {
                                    //triming extra spcaes
                                    let newOutputStream = ((self.outputStream ?? "") + output)
                                    if let range = newOutputStream.range(of: "\\S", options: .regularExpression) {
                                        self.outputStream = newOutputStream[range.lowerBound...].description
                                        self.chatHistory[self.chatHistory.count - 1].message = self.outputStream
                                        self.onOutputStreamUpdated?(self.outputStream)
                                    }
                                } else {
                                    self.outputStream = (self.outputStream ?? "") + output
                                    self.chatHistory[self.chatHistory.count - 1].message = self.outputStream
                                    self.onOutputStreamUpdated?(self.outputStream)
                                }

                            }
                        },
                        onFirstAudioGenerated: { [weak self] in
                            DispatchQueue.main.async {
                                self?.currentMessageLoading = false
                            }
                        },
                        onFinished: { [weak self] in
                            guard let self = self else { return }
                            
                            self.cancelLLM()
                            DispatchQueue.main.async {
                                self.cancelCurrentLLM()
                                self.isInterruptButtonVisible = false
                            }
                        },
                        onError: { [weak self] error in
                            print("LLM processing error: \(error)")
                            DispatchQueue.main.async {
                                self?.isLLMActive = false
                            }
                        }
                    )
                }
            }
        }

    }
    
    func addNewMessageToChatHistory(message: String, isUserInput: Bool) {
        
        if message.isEmpty { return }
        
        let newMessage = ChatMessage(
            message: message,
            isUserMessage: isUserInput,
            timestamp: Date(), tps: currentTPS ?? 0
        )
        
        chatHistory.append(newMessage)
        onNewMessageAdded?(newMessage)
        currentTPS = nil
        
        saveChatToRepository()
    }
    
    func saveChatToRepository() {
        guard let chatId = currentChatId else { return }
        let chat = Chat(messages: chatHistory, id: chatId)
        
        Task {
             cacheRepository.cacheChat(chat)
        }
    }
    
    func reset() {
        do {
            let res = NimbleNetApi.runMethod(methodName: "clear_prompt", inputs: [:])
            // saving current chat
            saveChatToRepository()
            resetUpdateCallBacks()
            cancelCurrentLLM()
            DispatchQueue.main.async { [weak self] in
                self?.chatHistory.removeAll()
            }
            createChatID()
            
        } catch {
            print("Error clearing chat: \(error.localizedMessage)")
        }
    }

    
    func resetUpdateCallBacks() {
        onNewMessageAdded = nil
        onChatHistoryUpdated = nil
        onOutputStreamUpdated = nil
        onLLMActive = nil
    }
    
    func createNewChat() {
        currentChatId = UUID().uuidString
    }
    
    deinit {
        cancelCurrentLLM()
        cancellables.forEach { $0.cancel() }
    }
}
